apps/designer/app/shared/tree-utils.ts

```diff
@@ -46,19 +53,131 @@ export const findSubtree = (
 };
 
 export const cloneInstance = (targetInstance: Instance) => {
-  const clonedIds = new Map<Instance["id"], Instance["id"]>();
+  const clonedInstanceIds = new Map<Instance["id"], Instance["id"]>();
   const clonedInstance = produce((targetInstance) => {
     const newId = ObjectId().toString();
-    clonedIds.set(targetInstance.id, newId);
+    clonedInstanceIds.set(targetInstance.id, newId);
     targetInstance.id = newId;
     traverseInstances(targetInstance, (instance) => {
       const newId = ObjectId().toString();
-      clonedIds.set(instance.id, newId);
+      clonedInstanceIds.set(instance.id, newId);
       instance.id = newId;
     });
   })(targetInstance);
   return {
-    clonedIds,
+    clonedInstanceIds,
     clonedInstance,
   };
 };
+
+export const cloneProps = (
+  props: Props,
+  clonedInstanceIds: Map<Instance["id"], Instance["id"]>
+) => {
+  const clonedProps: Props = [];
+  for (const prop of props) {
+    const instanceId = clonedInstanceIds.get(prop.instanceId);
+    if (instanceId === undefined) {
+      continue;
+    }
+    clonedProps.push({
+      ...prop,
+      id: ObjectId().toString(),
+      instanceId,
+    });
+  }
+  return clonedProps;
+};
+
+export const cloneStyleSources = (
+  styleSources: StyleSources,
+  subsetIds: Set<StyleSource["id"]>
+) => {
+  const clonedStyleSourceIds = new Map<Instance["id"], Instance["id"]>();
+  const clonedStyleSources: StyleSources = [];
+  for (const styleSource of styleSources) {
+    if (subsetIds.has(styleSource.id) === false) {
+      continue;
+    }
+    const newId = ObjectId().toString();
+    clonedStyleSourceIds.set(styleSource.id, newId);
+    clonedStyleSources.push({
+      ...styleSource,
+      id: newId,
+    });
+  }
+  return { clonedStyleSources, clonedStyleSourceIds };
+};
+
+export const cloneStyleSourceSelections = (
+  styleSourceSelections: StyleSourceSelections,
+  clonedInstanceIds: Map<Instance["id"], Instance["id"]>,
+  clonedStyleSourceIds: Map<Instance["id"], Instance["id"]>
+) => {
+  const clonedStyleSourceSelections: StyleSourceSelections = [];
+  for (const styleSourceSelection of styleSourceSelections) {
+    const instanceId = clonedInstanceIds.get(styleSourceSelection.instanceId);
+    if (instanceId === undefined) {
+      continue;
+    }
+    // preserve style source id when not cloned
+    // which means it is non-local style source
+    const values = styleSourceSelection.values.map(
+      (styleSourceId) =>
+        clonedStyleSourceIds.get(styleSourceId) ?? styleSourceId
+    );
+    clonedStyleSourceSelections.push({
+      values,
+      instanceId,
+    });
+  }
+  return clonedStyleSourceSelections;
+};
+
+// @todo migrate to style source variant
```

Comment:
do you have an issue or todo in an issue?

Reply:
This is for the next PR